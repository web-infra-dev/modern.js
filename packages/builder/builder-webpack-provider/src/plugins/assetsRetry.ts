import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
