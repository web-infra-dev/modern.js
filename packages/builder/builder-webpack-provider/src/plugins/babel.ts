import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
