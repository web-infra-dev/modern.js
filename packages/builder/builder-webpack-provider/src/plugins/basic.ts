import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
