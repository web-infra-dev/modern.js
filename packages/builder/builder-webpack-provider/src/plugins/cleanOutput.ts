import type { BuilderPlugin } from '../types';

export const PluginCleanOutput = (): BuilderPlugin => ({
  name: 'builder-plugin-clean-output',

  setup(api) {
    const clean = async () => {
      const config = api.getNormalizedConfig();

      if (config.output.cleanDistPath) {
        const { emptyDir } = await import('@modern-js/utils');
        const { distPath } = api.context;
        await emptyDir(distPath);
      }
    };

    api.onBeforeBuild(clean);
    api.onBeforeStartDevServer(clean);
  },
});
