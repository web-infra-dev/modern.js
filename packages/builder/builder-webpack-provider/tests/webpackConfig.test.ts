import { expect, describe, it } from 'vitest';
import { builderPluginBasic } from '@/plugins/basic';
import { createStubBuilder } from '@/stub';

describe('webpackConfig', () => {
  it('should allow tools.webpack to return config', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpack(config) {
            return {
              ...config,
              devtool: 'eval',
            };
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to modify config object', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpack(config) {
            config.devtool = 'eval';
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to be an object', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpack: {
            devtool: 'eval',
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to be an array', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpack: [
            {
              devtool: 'eval',
            },
            config => {
              config.devtool = 'source-map';
            },
          ],
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow to use tools.webpackChain to modify config', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpackChain(chain) {
            chain.devtool('eval');
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpackChain to be an array', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBasic()],
      builderConfig: {
        tools: {
          webpackChain: [
            chain => {
              chain.devtool('eval');
            },
            chain => {
              chain.devtool('source-map');
            },
          ],
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should export HtmlWebpackPlugin instance', async () => {
    await createStubBuilder({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            expect(utils.HtmlWebpackPlugin.version).toEqual(5);
          },
        },
      },
    });
  });

  it('should allow to append and prepend plugins', async () => {
    const builder = await createStubBuilder({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.appendPlugins([new utils.webpack.DefinePlugin({ foo: '1' })]);
            utils.prependPlugins([
              new utils.webpack.DefinePlugin({ foo: '2' }),
            ]);
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config.plugins).toMatchSnapshot();
  });

  it('should allow to remove plugins', async () => {
    const builder = await createStubBuilder({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.appendPlugins([new utils.webpack.DefinePlugin({ foo: '1' })]);
            utils.prependPlugins([
              new utils.webpack.DefinePlugin({ foo: '2' }),
            ]);
            utils.removePlugin('DefinePlugin');
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config.plugins).toEqual([]);
  });

  it('should allow to add rules', async () => {
    const newRule = {
      test: /\.foo$/,
      loader: 'foo-loader',
    };

    const builder = await createStubBuilder({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.addRules(newRule);
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config.module?.rules).toEqual([newRule]);
  });
});
