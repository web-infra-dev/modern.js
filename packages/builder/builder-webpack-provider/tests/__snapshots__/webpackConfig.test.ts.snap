// Vitest Snapshot v1

exports[`webpackConfig > should allow to append and prepend plugins 1`] = `
[
  DefinePlugin {
    "definitions": {
      "foo": "2",
    },
  },
  DefinePlugin {
    "definitions": {
      "foo": "1",
    },
  },
]
`;

exports[`webpackConfig > should allow to use tools.webpackChain to modify config 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to be an array 1`] = `
{
  "context": "<ROOT>",
  "devtool": "source-map",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to be an object 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to modify config object 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to return config 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpackChain to be an array 1`] = `
{
  "context": "<ROOT>",
  "devtool": "source-map",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should not have any pluginImport in Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /node_modules/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;

exports[`webpackConfig > should not set default pluginImport for Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /node_modules/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
              {
                "libraryDirectory": "es",
                "libraryName": "antd",
                "style": true,
              },
              "antd",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
              {
                "libraryDirectory": "es",
                "libraryName": "antd",
                "style": true,
              },
              "antd",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;

exports[`webpackConfig > should provide mergeConfig util in tools.webpack function 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should set proper pluginImport option in Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /node_modules/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
              {
                "camel2DashComponentName": true,
                "libraryName": "foo",
              },
              "foo",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-import/index.js",
              {
                "camel2DashComponentName": true,
                "libraryName": "foo",
              },
              "foo",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-app/src/babelPluginLockCorejsVersion",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-dynamic-import-node/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-lodash/index.js",
              {},
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-decorators/index.js",
              {
                "legacy": true,
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "helpers": false,
                "regenerator": true,
                "useESModules": true,
                "version": "7.21.5",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-export-default-from/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-pipeline-operator/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/@babel/plugin-proposal-partial-application/index.js",
            ],
            [
              "<WORKSPACE>/packages/cli/babel-preset-base/compiled/babel-plugin-styled-components/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
              "styled-components",
            ],
          ],
          "presets": [
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": false,
                "corejs": {
                  "proposals": true,
                  "version": "3.30",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": false,
                "shippedProposals": false,
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<WORKSPACE>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;
